File: core/src/main/java/dev/bee/kanjianki/core/study/HintProgression.java
package dev.bee.kanjianki.core.study;

public final class HintProgression {
    public HintVisibility visibility(HintState state, StrokeGuide guide) {
        HintState safeState = state == null ? HintState.initial() : state;
        int strokeCount = guide == null ? 0 : guide.strokeCount();
        int revealed = Math.min(safeState.revealedStrokeCount(), strokeCount);
        switch (safeState.level()) {
            case BLIND:
                return new HintVisibility(HintLevel.BLIND, false, false, false, false, false, revealed);
            case MINIMAL:
                return new HintVisibility(HintLevel.MINIMAL, false, false, false, true, true, Math.min(strokeCount, 1 + revealed));
            case OUTLINE:
                return new HintVisibility(HintLevel.OUTLINE, false, true, false, true, true, strokeCount);
            case TRACE:
            default:
                return new HintVisibility(HintLevel.TRACE, true, true, true, true, true, strokeCount);
        }
    }

    public HintState revealNext(HintState state, StrokeGuide guide) {
        HintState safeState = state == null ? HintState.initial() : state;
        int strokeCount = guide == null ? 0 : guide.strokeCount();
        if (safeState.level() == HintLevel.TRACE) {
            return new HintState(safeState.level(), strokeCount, 0);
        }
        if (safeState.level() == HintLevel.OUTLINE || strokeCount <= 0) {
            return new HintState(safeState.level().previous(), 0, 0);
        }
        if (safeState.revealedStrokeCount() >= strokeCount) {
            return new HintState(safeState.level().previous(), 0, 0);
        }
        return new HintState(
                safeState.level(),
                Math.min(strokeCount, safeState.revealedStrokeCount() + 1),
                0
        );
    }

    public HintState afterReview(HintState state, boolean writingPassed, int hintsUsed) {
        HintState safeState = state == null ? HintState.initial() : state;
        if (!writingPassed) {
            return new HintState(safeState.level().previous(), 0, 0);
        }
        if (hintsUsed > 0 || safeState.revealedStrokeCount() > 0) {
            return new HintState(safeState.level(), 0, 0);
        }
        return new HintState(safeState.level().next(), 0, safeState.consecutivePasses() + 1);
    }

    public HintState afterWriting(HintState state, WritingAnalysis analysis) {
        if (analysis == null) {
            return afterReview(state, false, 0);
        }
        if (analysis.status == WritingAnalysis.Status.CLOSE) {
            HintState safeState = state == null ? HintState.initial() : state;
            return new HintState(safeState.level(), 0, 0);
        }
        if (analysis.status != WritingAnalysis.Status.PASS) {
            return afterReview(state, false, analysis.hintsUsed());
        }
        return afterReview(state, true, analysis.hintsUsed());
    }
}

File: writing-core/src/main/kotlin/dev/bee/kanjianki/core/study/HintProgression.kt
package dev.bee.kanjianki.core.study

class HintProgression {
    fun visibility(state: HintState?, guide: StrokeGuide?): HintVisibility {
        val safeState = state ?: HintState.initial()
        val strokeCount = guide?.strokeCount() ?: 0
        val revealed = minOf(safeState.revealedStrokeCount(), strokeCount)
        return when (safeState.level()) {
            HintLevel.BLIND -> HintVisibility(HintLevel.BLIND, false, false, false, false, false, revealed)
            HintLevel.MINIMAL -> HintVisibility(
                HintLevel.MINIMAL,
                false,
                false,
                false,
                true,
                true,
                minOf(strokeCount, 1 + revealed),
            )

            HintLevel.OUTLINE -> HintVisibility(HintLevel.OUTLINE, false, true, false, true, true, strokeCount)
            HintLevel.TRACE -> HintVisibility(HintLevel.TRACE, true, true, true, true, true, strokeCount)
        }
    }

    fun revealNext(state: HintState?, guide: StrokeGuide?): HintState {
        val safeState = state ?: HintState.initial()
        val strokeCount = guide?.strokeCount() ?: 0
        if (safeState.level() == HintLevel.TRACE) {
            return HintState(safeState.level(), strokeCount, 0)
        }
        if (safeState.level() == HintLevel.OUTLINE || strokeCount <= 0) {
            return HintState(safeState.level().previous(), 0, 0)
        }
        if (safeState.revealedStrokeCount() >= strokeCount) {
            return HintState(safeState.level().previous(), 0, 0)
        }
        return HintState(
            safeState.level(),
            minOf(strokeCount, safeState.revealedStrokeCount() + 1),
            0,
        )
    }

    fun canRevealMoreHelp(state: HintState?, guide: StrokeGuide?): Boolean {
        if (state == null || state.level() == HintLevel.TRACE) {
            return false
        }
        return guide == null ||
            guide.isEmpty() ||
            state.level() == HintLevel.OUTLINE ||
            state.revealedStrokeCount() < guide.strokeCount()
    }

    fun afterReview(state: HintState?, writingPassed: Boolean, hintsUsed: Int): HintState {
        val safeState = state ?: HintState.initial()
        if (!writingPassed) {
            return HintState(safeState.level().previous(), 0, 0)
        }
        if (hintsUsed > 0 || safeState.revealedStrokeCount() > 0) {
            return HintState(safeState.level(), 0, 0)
        }
        return HintState(safeState.level().next(), 0, safeState.consecutivePasses() + 1)
    }

    fun afterWriting(state: HintState?, analysis: WritingAnalysis?): HintState {
        if (analysis == null) {
            return afterReview(state, false, 0)
        }
        if (analysis.status == WritingAnalysis.Status.CLOSE ||
            analysis.status == WritingAnalysis.Status.MODEL_UNAVAILABLE ||
            analysis.status == WritingAnalysis.Status.RECOGNITION_ERROR
        ) {
            val safeState = state ?: HintState.initial()
            return HintState(safeState.level(), 0, 0)
        }
        return afterReview(state, analysis.status == WritingAnalysis.Status.PASS, analysis.hintsUsed())
    }
}
